pkg/activator/revision_backends.go

```diff
@@ -0,0 +1,263 @@
+package activator
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"net/url"
+	"sync"
+	"time"
+
+	"go.uber.org/zap"
+	corev1 "k8s.io/api/core/v1"
+	corev1informers "k8s.io/client-go/informers/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"knative.dev/pkg/controller"
+	"knative.dev/serving/pkg/apis/networking"
+	"knative.dev/serving/pkg/apis/serving"
+	"knative.dev/serving/pkg/network"
+	"knative.dev/serving/pkg/network/prober"
+	"knative.dev/serving/pkg/queue"
+	"knative.dev/serving/pkg/reconciler"
+)
+
+type RevisionDestsUpdateFunc func(rev RevisionID, dests []string)
+
+const (
+	probePortName string        = "http"
+	probeTimeout  time.Duration = 300 * time.Millisecond
+)
+
+type revisionWatcher struct {
+	rev          RevisionID
+	updateCb     RevisionDestsUpdateFunc
+	dests        []string
+	healthyDests map[string]bool
+
+	transport http.RoundTripper
+	destsChan <-chan []string
+	logger    *zap.SugaredLogger
+}
+
+func newRevisionWatcher(rev RevisionID, updateCb RevisionDestsUpdateFunc,
+	destsChan <-chan []string, transport http.RoundTripper,
+	logger *zap.SugaredLogger) *revisionWatcher {
+	rw := revisionWatcher{
+		rev:          rev,
+		updateCb:     updateCb,
+		healthyDests: make(map[string]bool),
+		transport:    transport,
+		destsChan:    destsChan,
+		logger:       logger,
+	}
+	return &rw
+}
+
+func endpointsToDests(endpoints *corev1.Endpoints) []string {
+	ret := []string{}
+
+	for _, es := range endpoints.Subsets {
+		portVal := int32(0)
+		for _, port := range es.Ports {
+
+			if port.Name == probePortName {
+				portVal = port.Port
+				break
+			}
+		}
+
+		// This endpoint doesnt have a port we want to probe
+		// TODO(greghaynes) log error here
+		if portVal == 0 {
+			continue
+		}
+
+		for _, addr := range es.Addresses {
+			// Prefer IP as we can avoid a DNS lookup this way
+			ret = append(ret, fmt.Sprintf("%s:%d", addr.IP, portVal))
+		}
+	}
+
+	return ret
+}
+
+func filterHealthyDests(dests map[string]bool) []string {
+	ret := []string{}
+	for dest, healthy := range dests {
+		if healthy {
+			ret = append(ret, dest)
+		}
+	}
+	return ret
+}
+
+func (rw *revisionWatcher) checkDests() {
+	healthyDests := make(map[string]bool)
+	var healthyDestsMux sync.Mutex
+	var probeWg sync.WaitGroup
+
+	// Context used for our requests
+	ctx, cancel := context.WithTimeout(context.Background(), probeTimeout)
+	defer cancel()
+
+	for _, dest := range rw.dests {
+		// If the dest is already healthy then save this
+		if curHealthy, ok := rw.healthyDests[dest]; ok && curHealthy {
+			healthyDests[dest] = true
+		} else {
+			probeWg.Add(1)
+			go func(dest string) {
+				httpDest := url.URL{
+					Scheme: "http",
+					Host:   dest,
+				}
+				ok, err := prober.Do(ctx, rw.transport, httpDest.String(),
+					prober.WithHeader(network.ProbeHeaderName, queue.Name),
+					prober.ExpectsBody(queue.Name))
+
+				healthyDestsMux.Lock()
```

Comment:
if you put those into a channel and consume in a goroutine you can avoid this locking.